diff --git a/target/linux/ath79/dts/ar9331_tplink_tl-wr720n-v3.dts b/target/linux/ath79/dts/ar9331_tplink_tl-wr720n-v3.dts
new file mode 100644
index 0000000..e69de29
--- /dev/null
+++ b/target/linux/ath79/dts/ar9331_tplink_tl-wr720n-v3.dts
@@ -0,0 +1,96 @@
+/dts-v1/;
+#include "ar9331.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+
+/ {
+    compatible = "tplink,tl-wr720n-v3", "qca,ar9331";
+    model = "TP-Link TL-WR720N v3 (16M mod)";
+
+    aliases {
+        label-mac-device = &eth0;
+    };
+};
+
+&spi {
+    status = "okay";
+
+    flash@0 {
+        compatible = "jedec,spi-nor";
+        reg = <0>;
+
+        partitions {
+            compatible = "fixed-partitions";
+            #address-cells = <1>;
+            #size-cells = <1>;
+
+            partition@0 {
+                label = "u-boot";
+                reg = <0x000000 0x020000>;
+                read-only;
+            };
+
+            partition@20000 {
+                label = "u-boot-env";
+                reg = <0x020000 0x010000>;
+            };
+
+            partition@30000 {
+                label = "firmware";
+                reg = <0x030000 0xfc0000>;
+            };
+
+            art: partition@ff0000 {
+                label = "art";
+                reg = <0xff0000 0x010000>;
+                read-only;
+
+                nvmem-layout {
+                    compatible = "fixed-layout";
+                    #address-cells = <1>;
+                    #size-cells = <1>;
+
+                    macaddr_art_0: macaddr@0 {
+                        reg = <0x0 0x6>;
+                    };
+                };
+            };
+        };
+    };
+};
+
+&eth0 {
+    status = "okay";
+    nvmem-cells = <&macaddr_art_0>;
+    nvmem-cell-names = "mac-address";
+};
+
+&wmac {
+    status = "okay";
+    mtd-cal-data = <&art 0x1000>;
+
+    nvmem-cells = <&macaddr_art_0>;
+    nvmem-cell-names = "mac-address";
+};
+
diff --git a/target/linux/ath79/image/tiny-tp-link.mk b/target/linux/ath79/image/tiny-tp-link.mk
index 0000000..0000000 100644
--- a/target/linux/ath79/image/tiny-tp-link.mk
+++ b/target/linux/ath79/image/tiny-tp-link.mk
@@ -1,3 +1,12 @@
+
+define Device/tl-wr720n-v3
+  $(Device/tplink-8mlzma)
+  SOC := ar9331
+  DEVICE_TITLE := TP-Link TL-WR720N v3 (16M mod)
+  DEVICE_PACKAGES := kmod-usb-core kmod-usb2
+endef
+TARGET_DEVICES += tl-wr720n-v3
+
